File: common/src/main/java/juuxel/adorn/block/entity/AdornBlockEntityType.java
package juuxel.adorn.block.entity;

import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntityType.BlockEntitySupplier;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;

import java.util.Set;
import java.util.function.Predicate;

public final class AdornBlockEntityType<E extends BlockEntity> extends BlockEntityType<E> {
    private final Predicate<Block> blockPredicate;

    public AdornBlockEntityType(BlockEntitySupplier<? extends E> factory, Predicate<Block> blockPredicate) {
        super(factory, Set.of());
        this.blockPredicate = blockPredicate;
    }

    @Override
    public boolean isValid(BlockState state) {
        return blockPredicate.test(state.getBlock());
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/entity/AdornBlockEntityType.kt
package juuxel.adorn.block.entity

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.util.math.BlockPos

class AdornBlockEntityType<E : BlockEntity>(
    factory: (BlockPos, BlockState) -> E,
    private val blockPredicate: (Block) -> Boolean
) : BlockEntityType<E>(factory, emptySet(), null) {
    override fun supports(state: BlockState) = blockPredicate(state.block)
}
